add decay

Patrick-Lascombe opened the pull request:
Adding an optional decay on created points

Patrick-Lascombe committed:
add decay

DLu commented on range_sensor_layer/include/range_sensor_layer/range_sensor_layer.h:
```diff
@@ -65,6 +67,7 @@ class RangeSensorLayer : public costmap_2d::CostmapLayer
   boost::function<void(sensor_msgs::Range& range_message)> processRangeMessageFunc_;
   boost::mutex range_message_mutex_;
   std::list<sensor_msgs::Range> range_msgs_buffer_;
+  std::map<std::pair<int, int>, double> point_map;
```
A) Please use the trailing underscore format for private variables
B) Did you choose signed ints over unsigned ints for a reason?
C) From the variable name, it is unclear what the values represent.

Patrick-Lascombe replied:
Concerning the B, I used int instead of unsigned int by habit. But since we use index that are only positive we could used unsigned int instead of int for the map

DLu commented on range_sensor_layer/src/range_sensor_layer.cpp:
```diff
@@ -250,7 +253,9 @@ void RangeSensorLayer::updateCostmap(sensor_msgs::Range& range_message, bool cle
-  if (!tf_->canTransform(global_frame_, in.header.frame_id, in.header.stamp, ros::Duration(0.1)))
+  std::map<int, int>::iterator it_pair;
+
+  if (!tf_->canTransform(global_frame_, in.header.frame_id, in.header.stamp, ros::Duration(0.3)))
```
Can you parameterize the fourth parameter here?

DLu commented on range_sensor_layer/src/range_sensor_layer.cpp:
```diff
@@ -250,7 +253,9 @@ void RangeSensorLayer::updateCostmap(sensor_msgs::Range& range_message, bool cle
   in.header.frame_id = range_message.header.frame_id;
 
-  if (!tf_->canTransform(global_frame_, in.header.frame_id, in.header.stamp, ros::Duration(0.1)))
+  std::map<int, int>::iterator it_pair;
```
This iterator is not used.

DLu commented on range_sensor_layer/src/range_sensor_layer.cpp:
```diff
@@ -354,6 +359,21 @@ void RangeSensorLayer::updateCostmap(sensor_msgs::Range& range_message, bool cle
+  std::map<std::pair<int,int>, double>::iterator it_map;
+  for (it_map = point_map.begin() ; it_map != point_map.end() ; it_map++ )
+  {
+    if(pixel_decay_ < last_reading_time_.toSec() - it_map->second)
```
Just for minor efficiency, can you move the conversion of last_reading_time_ to a double outside of the loop?

DLu commented on range_sensor_layer/src/range_sensor_layer.cpp:
```diff
@@ -376,7 +396,20 @@ void RangeSensorLayer::update_cell(double ox, double oy, double ot, double r, do
+    if(use_decay_)
+    {
+      if(c > to_cost(mark_threshold_))
+        point_map[std::make_pair(x, y)] = last_reading_time_.toSec();
```
Since you use the pair(x,y) in all branches of this, can you declare it once inside the `use_decay` if statement?

DLu reviewed (changes requested):
Thanks for this contribution. Couple small changes requested. Bonus points if you want to write a test.

Patrick-Lascombe committed:
Changing name of var point_map

point_map is now marked_point_history to be more clear and respect the format

Patrick-Lascombe committed:
Parametrisation of the transform waiting time

And deletion of useless iterator

Patrick-Lascombe committed:
Code optimisation

Declaration of last_reading_time_sec outside of for loop and of coordinate pair outside of if loop

Patrick-Lascombe committed:
Small correction

DLu commented on range_sensor_layer/include/range_sensor_layer/range_sensor_layer.h:
```diff
@@ -79,6 +82,10 @@ class RangeSensorLayer : public costmap_2d::CostmapLayer
 
+  bool use_decay_;
+  double pixel_decay_;
+  double US_transform_waiting_time_;
```
What is US here?

Patrick-Lascombe replied:
Ultra Sound, I used US to make a shorter variable name

DLu replied:
Since the sensor isn't assumed to be an ultrasound, can we just make it the more generic `transform_tolerance_`? Standard is not to use capital letters anyway.

DLu commented on range_sensor_layer/include/range_sensor_layer/range_sensor_layer.h:
```diff
@@ -45,11 +45,13 @@ class RangeSensorLayer : public costmap_2d::CostmapLayer
 
   void updateCostmap();
   void updateCostmap(sensor_msgs::Range& range_message, bool clear_sensor_cone);
+  void timeCheck();
```
Would you mind renaming this to `removeOutdatedReadings()`

DLu commented on range_sensor_layer/include/range_sensor_layer/range_sensor_layer.h:
```diff
@@ -45,11 +45,13 @@ class RangeSensorLayer : public costmap_2d::CostmapLayer
   double delta(double phi);
   double sensor_model(double r, double phi, double theta);
 
+
```
Unnecessary blank line

DLu commented on range_sensor_layer/include/range_sensor_layer/range_sensor_layer.h:
```diff
@@ -65,6 +67,7 @@ class RangeSensorLayer : public costmap_2d::CostmapLayer
   boost::function<void(sensor_msgs::Range& range_message)> processRangeMessageFunc_;
   boost::mutex range_message_mutex_;
   std::list<sensor_msgs::Range> range_msgs_buffer_;
+  std::map<std::pair<int, int>, double> marked_point_history_;
```
I still feel the correct thing is to use `unsigned int` for this.

DLu commented on range_sensor_layer/src/range_sensor_layer.cpp:
```diff
@@ -354,6 +358,23 @@ void RangeSensorLayer::updateCostmap(sensor_msgs::Range& range_message, bool cle
+
+  for (it_map = marked_point_history_.begin() ; it_map != marked_point_history_.end() ; it_map++ )
+  {
+    if(pixel_decay_ < last_reading_time_sec - it_map->second)
```
If we really wanted to optimize, we could calculate `removal_time = last_reading_time.toSec() - pixel_decay_` and then simply check if `it_map->second < removal_time`

DLu commented on range_sensor_layer/src/range_sensor_layer.cpp:
```diff
@@ -376,7 +397,21 @@ void RangeSensorLayer::update_cell(double ox, double oy, double ot, double r, do
     setCost(x, y, c);
+    if(use_decay_)
+    {
+      std::pair<int, int> coordinate_pair(x, y);
```
Can you please add a comment to this section to explain what is going on (particularly with the else if )

DLu reviewed (changes requested):
I apologize for the glacial pace of my PR reviews.

Patrick-Lascombe committed:
Renaming and small changes

Patrick-Lascombe committed:
Pass map point to unsigned int

Patrick-Lascombe committed:
Optimise removal_time check

Patrick-Lascombe committed:
Last passation to unsigned int

Patrick-Lascombe committed:
Adding comments

Patrick-Lascombe commented:
Sorry, this time it was me that let this PR sit for a while. Here are the requested changes.
If you need any other changes do not hesitate !

doisyg committed:
Merge pull request #1 from DLu/melodic

upsteam changes

DLu commented:
Cherry-picked to noetic too. 

Again, I apologize for the glacial pace.
